tracing: Add OpenTracing support

jodh-intel opened the pull request:
Add OpenTracing [1] support using Jaeger [2]. Full details are provided
in `TRACING.md`.

Fixes #322.

[1] - https://opentracing.io
[2] - https://jaegertracing.io

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

jodh-intel commented:
Related: #328.

klynnrif commented on TRACING.md:
```diff
@@ -0,0 +1,312 @@
+  [Guest to Host communication using VSOCK](#guest-to-host-communication-using-vsock)
+  section for further details.
+
+- Tracing is only finalised when the workload has exited
```
finalised => finalized

klynnrif commented on TRACING.md:
```diff
@@ -0,0 +1,312 @@
+
+- Tracing is only finalised when the workload has exited
+
+  Although trace *can* be inspected before the workload exits, it will be
```
Suggested rewrite to stay active: Although trace *can* be inspected before the workload exits, it is incomplete.

klynnrif commented on TRACING.md:
```diff
@@ -0,0 +1,312 @@
+
+# Appendix: Setting up a tracing development environment
+
+If you wish to debug, further develop or test tracing, follow the steps below
```
Add series comma, wish=>want, and below=>following: 

If you want to debug, further develop, or test tracing, use the following steps...

klynnrif commented on TRACING.md:
```diff
@@ -0,0 +1,312 @@
+## Guest journal redirector
+
+Since agent tracing requires VSOCK, and enabling VSOCK in `configuration.toml`
+disables the `kata-proxy` (since it is not required), the Kata Container has
```
Lines 291-294 suggested rewrite: disables the 'kata-proxy' (not required), the Kata Container has no way to display the guest OS journal messages, which are normally redirected through the proxy. This makes debugging difficult. To overcome this issue, create another pair of ...

devimc commented on agent.go:
```diff
@@ -352,13 +425,22 @@ func (s *sandbox) getContainer(id string) (*container, error) {
 
-func (s *sandbox) setContainer(id string, ctr *container) {
+func (s *sandbox) setContainer(ctx context.Context, id string, ctr *container) {
+	span, _ := trace(ctx, "sandbox", "setContainer")
```
other functions use `s.trace`, why don't use it here?

devimc commented on agent.go:
```diff
@@ -224,6 +255,19 @@ func (c *container) getProcess(execID string) (*process, error) {
 
+func (s *sandbox) trace(name string) (opentracing.Span, context.Context) {
+	if s.ctx == nil {
+		agentLog.WithField("type", "bug").Error("trace called before context set")
```
might be this function should be a `singleton`, hence log an error is not needed

jodh-intel replied:
I added this for consistency with the approach used in the runtime. If you look at `agent.go`, you'll see that there is a "gap" between creating the sandbox and having `s.ctx` set. That is unfortunately required since:

- the context can only be set after tracing has been setup.
- the sandbox needs to be created before tracing has been setup.

We could use a new context when creating the sandbox and then reset it to the correct context containing the trace metadata but this approach is simpler I think.

devimc commented:
nice job James! :smile:

jshachm commented:
Nice job @jodh-intel 

Beyond this, will we have a chance to support this in `initrd` model

jodh-intel commented:
Thanks @klynnrif, @devimc - branch updated.

@sboeuf - hahaaa! It's not as bad as it might appear in terms of LOC. But it is a little subtle in places, hence the big `TRACING.md`.

jodh-intel commented:
Added dnm to make it clearer there is a block on the agent PR.

jodh-intel commented:
Note that the horrid hex strings in the scripts and Jaeger systemd services are going to be architecture-dependent. That's not a problem today but probably will be once os390x support lands. It's unclear how best to handle this. A reliable mechanism would be to provide a C program to generate the output as @mcastelino does on https://gist.github.com/mcastelino/9a57d00ccf245b98de2129f0efe39857#using-specific-ports, but the ideal would be to not require users to build and run a C program simply to get tracing working on big endian systems.

jodh-intel commented on Makefile:
```diff
@@ -15,6 +15,16 @@ PREFIX := /usr
+# Set to "yes" if agent should support OpenTracing with http://jaegertracing.io.
+TRACE := no
+
+ifeq ($(TRACE),yes)
```
Good point - fixed.

jodh-intel commented on Makefile:
```diff
@@ -25,9 +25,17 @@ ifeq ($(TRACE),yes)
 endif
 
+# If "yes", install additional services to redirect guest OS journal messages
+# to the host using VHOST.
```
Fixed.

jodh-intel commented on Gopkg.toml:
```diff
@@ -50,3 +50,7 @@
+
+[[constraint]]
+  name = "github.com/uber/jaeger-client-go"
+  version = "2.15.0"
```
That was auto-added by `dep`, but I agree it makes sense so fixed.

jodh-intel commented on TRACING.md:
```diff
@@ -0,0 +1,312 @@
+- A pair of timestamps (recording the start time and end time of some operation)
+- A reference to the span's parent span
+
+All spans need to be "finished", or completed, to allow the OpenTracing
```
Good point - fixed.

jodh-intel commented on TRACING.md:
```diff
@@ -0,0 +1,312 @@
+
+- Adding a Jaeger agent process into the VM would require
+  special-case code for handling running the `kata-agent`
+  [as the init daemon](https://github.com/kata-containers/osbuilder).
```
np ;)

jodh-intel commented on TRACING.md:
```diff
@@ -0,0 +1,312 @@
+
+# VM shutdown
+
+Assume a user has run the following to create a Kata Container:
```
Agreed - fixed.

jodh-intel commented on TRACING.md:
```diff
@@ -0,0 +1,312 @@
+
+  - The `kata-agent` `DestroySandbox` gRPC API is called.
+  - The runtime sends a [QMP](https://wiki.qemu.org/Documentation/QMP) `quit`
+    command to the VM to ensure it is stopped.
```
Fixed.

jodh-intel commented on TRACING.md:
```diff
@@ -0,0 +1,312 @@
+  The user exits the primary workload shell by typing `exit`.
+
+  This results in the `kata-agent` process detecting that the primary workload
+  process has exited, and then exits itself. This, in turn, causes the VM to shut down.
```
I did indeed do this in testing but due I suspect to the containerd issue, that change wasn't required.

This will no doubt need updating as you suggest but let me go and look at the containerd version...

jodh-intel replied:
Picking this up again...

We don't want to *stop* the kata agent service stopping the VM - it's the other way around for tracing scenarios:

- If the user types `exit`:
  - The workload shell exits
  - The agent detects this and stops tracing itself.
  - The agent exits.
  - The agent service detects this and performs a controlled VM shutdown.

- If the user types `docker stop`, the same thing has to happen - the VM needs to shut itself down and the runtime should not call the QMP `QUIT` command as that will kill the VM before the agent has finished stopping tracing itself.

That does mean I'll have to tweak the runtime "somehow" to stop it calling `QUIT`. I can't see currently how to make that change "clean" - currently, you enable agent trace by modifying `hypervisor.qemu.kernel_params` directly. The best I can think of right now is to add a `runtime.enable_agent_tracing` which will:

- Disable the `QUIT`.
- Add the trace arguments to `hypervisor.qemu.kernel_params` (yuck).

Any other ideas?

sboeuf replied:
I don't get what's wrong with the gRPC call `DestroySandbox()`. This function is always called in every situation (now that we fixed Docker 18.09), meaning the agent is always around, even when the container workload exited.
In both cases (container workload exit, and forced kill), the runtime will at some point communicate with the agent the need to terminate through `agent.stopSandbox()` which will send the gRPC call `DestroySandbox()`.

jodh-intel replied:
Yes - `DestroySandbox()` is always called, but it doesn't appear to be possible to shut down the gRPC server by triggering the server to end from `DestroySandbox()`. I need to debug further but I think this is the same problem I hit a while back with gRPC internal locks and waitgroups. It's sounding like you cannot shut down the server on the same goroutine that's it's running in. However, if you try to stop the server in `WaitProcess()`, all works as expected.

jodh-intel commented on TRACING.md:
```diff
@@ -0,0 +1,312 @@
+The crucial difference between the two shutdown methods explained in the [VM
+shutdown](#vm-shutdown) section is that the user-initiated shutdown does not
+explicitly destroy the sandbox using the `DestroySandbox` gRPC API: the
+sandbox is "destroyed" by the agent shutting down the VM.
```
Again, this relates to the issue I seem to have been having with containerd so I'll come back to this one...

grahamwhaley commented:
@sboeuf - this probably needs your re-review now when you find a slot.

grahamwhaley commented on TRACING.md:
```diff
@@ -0,0 +1,312 @@
+  * [Host environment](#host-environment)
+  * [Guest environment](#guest-environment)
+* [Limitations](#limitations)
+* [Appendix: Setting up a tracing development environment](#appendix:-setting-up-a-tracing-development-environment)
```
The appendix link does not hot-render. I think it needs the ref modding to (that is, drop the stray `:`) :
```
#appendix-setting-up-a-tracing-development-environment
```

jodh-intel replied:
Fixed.

grahamwhaley commented on TRACING.md:
```diff
@@ -0,0 +1,312 @@
+following:
+
+| Trace type | Trace setting | Description | Use-case | Notes |
+|-|-|-|-|
```
This table is not rendering for me - I suspect you are missing one set of `| - |` in the second line.

jodh-intel replied:
Weirdly it rendered for me in both github and atom. What are you using ooi?

grahamwhaley replied:
I was just looking in the 'view file' on the github PR - so, dunno why that was different from your github view...

grahamwhaley commented on TRACING.md:
```diff
@@ -0,0 +1,312 @@
+
+> **Note:**
+>
+> Agent tracing is separate from the tracing for other Kata Containers
```
I have a curiosity question - I presume you have traced the agent alongside all the other components, and all the trace items turn up in the right order in the view? I ask as sometimes we get asked about time skew in the VM, and if that existed then maybe it would mess up the trace (presuming Jaeger uses the timestamps, and not the order the messages turn up at the server to sort the events...)

jodh-intel replied:
afaik, the tracing is correlated in terms of timestamps and span contexts. I haven't seen any issues with "collated tracing mode" (which I think is what you're implicitly referring to), but I'd appreciate it if as many folk as possible could test this PR to further validate ;)

grahamwhaley commented on TRACING.md:
```diff
@@ -0,0 +1,312 @@
+1. Build a custom non-initrd image using
+   [osbuilder](https://github.com/kata-containers/osbuilder).
+
+   - The image **must** include:
```
How does one achieve this? I think you added the ability to osbuilder to do this for you?, in which case we can note the method (TRACE=true or somesuch) here?

jodh-intel replied:
See https://github.com/kata-containers/osbuilder/pull/200 :)

It's a bit circular. Once the osbuilder PR has landed, this can be simplified.

I did toy with the idea of creating a small `.md` doc here and having osbuilder execute it, but discarded the idea since that script needs to know things like `$ROOTFS` directories and those are a purely osbuilder concern so it would have been "the wrong way round".

grahamwhaley commented on TRACING.md:
```diff
@@ -0,0 +1,312 @@
+1. Install and run Jaeger on the host.
+
+   The simplest method to start using Jaeger is to follow the instructions to run the
+   ["all-in-one" Docker\* image](https://www.jaegertracing.io/docs/1.8/getting-started/#all-in-one).
```
Heh, side question then - does that Jaeger all-in-one docker image work with kata as the runtime?? :-)

jodh-intel replied:
no ;(

jodh-intel replied:
Recorded as https://github.com/kata-containers/runtime/issues/942.

jodh-intel commented on grpc.go:
```diff
@@ -806,6 +807,10 @@ func (a *agentGRPC) SignalProcess(ctx context.Context, req *pb.SignalProcessRequ
 	if err != nil {
+		// primary process no longer exists so trigger a shutdown of
+		// the gRPC server and thus the agent.
+		a.requestServerStop()
```
@sboeuf - `requestServerStop()` seems to *have* to be called here: moving it to `DestroySandbox()` does not dtrt and stops the gRPC server from ending. By having the call here, we're waiting for the shim to call `WaitProcess()`.

What I don't particularly like is the "stop gRPC server on any error from `getProcess()`" approach (similar to https://github.com/kata-containers/runtime/pull/940). I could determine if it was a grpc `NotFound` error I guess?

sboeuf replied:
@jodh-intel 
I'm sorry but now that I took a detailed look at the entire code, I think that calling `a.requestServerStop()` from there is not the right thing to do.

`WaitProcess()` is the function that is called by every shim on the host, representing a process. It could be any container process, or any exec'ed process running into an existing container.
Based on this, it does not seem appropriate to ask the server to be stopped on any `WaitProcess()` error. It should be obvious from the code, that we request the server to be stopped when the sandbox is being stopped and removed, or from any piece of code that knows for sure that the agent is done and about to exit.

Based on the current code, I don't understand how a properly running container will end up calling `requestServerStop()` since I don't expect any error to come out of `getProcess()`.

And in case of a multi containers pod, if one container process goes wrong for some reason, do we want to force the grpc server to be stopped, and the agent at the same time?

jodh-intel commented:
Ping @sboeuf et @kata-containers/agent - tasty new code y'all! ;)

sboeuf commented:
/test

jodh-intel commented:
/test

jodh-intel commented:
/retest

jodh-intel commented:
/retest

jodh-intel commented:
@sboeuf - doc updated. I've removed the offending shutdown details and added more info to the comment for the `a.requestServerStop()` call.

jodh-intel commented:
@egernst - ptal.

sboeuf commented on agent.go:
```diff
@@ -98,17 +103,22 @@ type sandboxStorage struct {
+	server      *grpc.Server
+
+	// Set when server needs to be shut down
+	shutdown chan bool
```
Instead you could use:
```go
shutdown chan struct{}
```
since you don't care about the boolean value anywhere in the code.

sboeuf commented on grpc.go:
```diff
@@ -1255,6 +1266,33 @@ func (a *agentGRPC) CreateSandbox(ctx context.Context, req *pb.CreateSandboxRequ
+	}
+
+	// Inform the caller the server should end
+	a.sandbox.shutdown <- true
```
Not using a boolean Go channel, but a struct{} instead would be more appropriate. This would lead to modify this line using:
```go
close(a.sandbox.shutdown)
```

egernst commented:
@jodh-intel @sboeuf 

I tested a failing to launch container with this PR included in the agent. With a standard agent, it would fail pretty quick.  I noticed that the failing docker run command would sit for a long time (minutes) when using this PR as is.  I then applied a change suggested by @sboeuf  at https://gist.githubusercontent.com/sboeuf/574b8e473142b94cf36ad4091d1f2321/raw/4d5e78ad897c806bfe7dad161553623f668a4313/agent_patch.patch on top of this PR, and I see expected behaviour (just a quick failure and return).

Hope this helps clarify/verify comment: https://github.com/kata-containers/agent/pull/415#discussion_r239617539

jodh-intel commented:
Hi @sboeuf - I've had limited time to look at this today but adding some debug shows that there are two calls to `WaitProcess()` - the shim and the runtime. The runtime hits the error scenario where `a.sandbox.getProcess()` returns an error in `WaitProcess()` of type: `rpc error: code = NotFound`.

I'll take another look at what is blocking `DestroySandbox()` from triggering the server shutdown tomorrow though...

sboeuf commented:
@jodh-intel any chance you could spend some time debugging this?

jodh-intel commented on jaeger-client-socat-redirector.service:
```diff
@@ -0,0 +1,45 @@
+#
+# - https://gist.github.com/mcastelino/9a57d00ccf245b98de2129f0efe39857#using-specific-ports
+# - http://www.dest-unreach.org/socat/doc/socat-genericsocket.html
+Environment=host_vsock_addr=x00x00xafx1ax00x00x02x00x00x00x00x00x00x00
```
This is fragile as it encodes a `sockaddr_vm` structure in hex form.

[I don't see any sign](https://repo.or.cz/socat.git/shortlog) that `socat` is being updated to add "native" vsock support so we need to either look at nudging that along or find a workaround that improves on what I have here.

jodh-intel commented:
Blocked on #445.

jodh-intel commented:
This branch builds on @sboeuf's fix on #448. It *mostly* works but there are a few things to note:

- If you haven't been following along, you **must** build the agent with golang 1.11* or newer to ensure correct behaviour.

- Currently, only isolated tracing is working (due to https://github.com/kata-containers/runtime/issues/1277).

- I've retained @sboeuf's commit to disable VM shutdown on agent exit but this is a problem...

  With that behaviour plus the existing runtime behaviour where it sends the QMP QUIT, the runtime still nukes the agent before it's finished reporting spans.

  Either the runtime needs to wait until the agent has fully exited before it nukes it, or the agent needs to somehow report that it's dead. Neither approach is palatable as the gRPC server is no longer running at this point and having the runtime poll the server until it no longer responds is the road to madness imho.

  The above is the reason my original approach was to have the agent shut down the VM as it's simple and reliable and the runtime doesn't need to be involved.

grahamwhaley commented:
Time to bump our [version of golang in the versions.yaml file](https://github.com/kata-containers/runtime/blob/master/versions.yaml#L215) yet? We currently have:
- 1.10.4 listed as our prime version
- 1.11.1 as our newest tested
- but 1.11.5 is out
- and 1.10 is up to 1.10.8 now...

jodh-intel commented:
@grahamwhaley - I think so yes ;)

sboeuf commented:
@jodh-intel 
>    With that behaviour plus the existing runtime behaviour where it sends the QMP QUIT, the runtime still nukes the agent before it's finished reporting spans.
      Either the runtime needs to wait until the agent has fully exited before it nukes it, or the agent needs to somehow report that it's dead. Neither approach is palatable as the gRPC server is no longer running at this point and having the runtime poll the server until it no longer responds is the road to madness imho.
      The above is the reason my original approach was to have the agent shut down the VM as it's simple and reliable and the runtime doesn't need to be involved.

I've put more thoughts into this and I think I have to agree that letting the agent properly ending should be part of Kata. So here are my suggestions:
- we will assume that the VM will be shut down from the agent
- we will need to provide the agent with the appropriate way of killing the VM. Just want to make sure that shutting down the VM is not an implicit thing related to systemd `kata-agent.service` description. This will ensure the agent being in control and actually knowing what it is doing, instead of assuming that because the agent goes away, the VM goes away.
- from the runtime, we will have to wait for the end of the VM. If we don't do that, we might end up reusing the same VM sockets path between two pods with the same pod ID, because it might take some time to shutdown the first VM while we're already starting up the second one.
- from the runtime, we will need to add a timeout to our wait for the VM to exit, so that if it takes more than 10seconds (we need to agree on a number here), we want the runtime to explicitly kill the VM. In case of QEMU this would mean `QMP_QUIT`, in case of Firecracker, this would mean `SendCtrlAltDel`.

If we're going through those 4 steps, we should be able to maintain the same behavior we have right now, but also be able to let the agent terminate correctly, which would give it the time to send the tracing back to the host.
@jodh-intel WDYT about this proposal?

One last point, we need to test all that with non-vsock use case (proxy basically) since I think we will have a few issues with the last gRPC call...

jodh-intel commented:
osbuilder PR https://github.com/kata-containers/osbuilder/pull/200 updated for latest changes here.

jodh-intel commented:
@sboeuf - If we make the agent explicitly shutdown the VM, we might end up back in a"catch 22" situation - if the agent calls something like `shutdown(8)`, the agent must be running. But if the agent is running it might still have outstanding trace spans. That said, looking at the code on this PR, we're "safe" in that we can arrange for the shutdown to happen in `main()`. That should be acceptable since the tracing ends in `realMain()`. That is, assuming the UDP trace packets are immediately routed out of the VM. We'll need a clear comment to avoid anyone creating trace spans that cannot be finalised due to the VM shutdown.

Regarding who shuts down the VM, from the host perspective, we're talking about the difference between these two scenarios:

- Runtime calls `DestroySandbox()` and waits for **systemd** to shutdown VM, killing the VM if the shutdown takes too long.

- Runtime calls `DestroySandbox()` and waits for **the Kata agemt** to shutdown VM, killing the VM if the shutdown takes too long.

Currently, we don't wait for the VM to end, but I agree that it would be safer to do so in this scenario. Any thoughts on the best way to do that with govmm @markdryan?

There may be security and performance implications of the decisions we make here.

One final point: it would be pretty awesome to get agent tracing landed in a basic form before the epoch rollover, so if we can find a safe subset of these extra changes, that would be helpful I think as we could start tracing and refine the approach over time potentially ;)

/cc @sameo, @mcastelino, @markdryan, @egernst, @gnawux, @bergwolf.

sboeuf commented:
@jodh-intel 
>Currently, we don't wait for the VM to end, but I agree that it would be safer to do so in this scenario.

Well we don't explicetly, but we do because `QMP_QUIT` is blocking until the VM is properly killed.

jodh-intel commented:
Awaiting feedback from a bunch of people... :smile:

jodh-intel commented:
Re-ping @sameo, @mcastelino, @markdryan, @egernst, @gnawux, @bergwolf.

grahamwhaley commented on kata-agent.service.in:
```diff
@@ -1,3 +1,9 @@
+#
+# Copyright (c) 2018 Intel Corporation
```
2019

jodh-intel replied:
Fixing...

grahamwhaley commented on kata-agent.service.in:
```diff
@@ -10,5 +16,4 @@ StandardOutput=tty
 LimitNOFILE=infinity
-ExecStop=/bin/sync ; /usr/bin/systemctl --force poweroff
-FailureAction=poweroff
+ExecStop=/bin/sync
```
do we even need the sync? I guess maybe we are trying to flush fs buffers out of the VM back to the host?

amshinde replied:
@grahamwhaley Yes, we need the sync for block based storage drivers to flush the buffers to the block device.

grahamwhaley commented on README.md:
```diff
@@ -15,3 +16,8 @@ The agent manages container processes inside the VM, on behalf of the
+## Developer mode
+
+Add `agent.devmode` to the guest kernel command line to enable
+[debug mode](#debug-mode) and allow the agent process to coredump (disabled by default).
```
are you saying the above debug mode relies on this developer mode to work? In which case, maybe I'd put the two sections the other way around, or combine them. As it stands, I'm a bit confused about what each mode (debug and dev) gets you, and what their relationship is.

jodh-intel replied:
I've reworded this slightly but the details are there if you squint ;) Debug mode enabled debug output (same for all Kata components), whereas developer mode enabled coredumping (but also implicitly enables debug mode to save you having to specify both options).

jodh-intel replied:
... and before anyone comments, yes, I'm still updating this doc! ;)

grahamwhaley commented on .gitignore:
```diff
@@ -1,4 +1,3 @@
 kata-agent
 kata-agent.service
-agent
```
maybe this should be a `/agent`, to just ignore the agent file in the top level dir?
tbh, I don't immediately see how it would be affecting anything under `vendor`, unless there were a file just called `agent` in there - but, prefixing with `/` I'm thinking might be the right thing to do?

jodh-intel replied:
As a general rule, we should be ignoring as few files as possible. But that entry was causing problems as it is bogus (now). It may have been valid once but retaining it causes problems for files such as:

```
vendor/github.com/uber/jaeger-client-go/thrift-gen/agent/agent.go
```

grahamwhaley replied:
right - adding the `/` on the front should make it say 'just the agent file in the top level dir' afaict - which feels like what the original intent was.

jodh-intel replied:
That makes sense. But we don't need that rule any more ;)

grahamwhaley commented on TRACING.md:
```diff
@@ -0,0 +1,272 @@
+Further details:
+
+| Name | Type | Environment | Summary | Details | Explanation |
+|-|-|-|-|-|
```
I think you are missing one ` - |` here, which is preventing the table from rendering.

grahamwhaley commented on TRACING.md:
```diff
@@ -0,0 +1,272 @@
+# Running the agent with tracing enabled
+
+1. Build a custom non-initrd image using
+   [osbuilder](https://github.com/kata-containers/osbuilder).
```
Do we/can we have a better reference here - specifically stating to set `TRACE=true` in the osbuilder env/makefile for instance?

jodh-intel replied:
This PR is "purely" for the agent. For what you're after, you prolly want to check out https://github.com/kata-containers/osbuilder/pull/200.

grahamwhaley commented on TRACING.md:
```diff
@@ -0,0 +1,272 @@
+   >
+   > Currently, it is necessary to run the `all-in-one` image with `--runtime runc`.
+   >
+   > See: https://github.com/kata-containers/runtime/issues/942.
```
issue 942 appears to be closed as 'not a problem' - do you need to update this item?

jodh-intel replied:
Dropping...

mcastelino commented on TRACING.md:
```diff
@@ -0,0 +1,272 @@
+# Introduction
+
+This document explains agent tracing, which uses 
+the [Jaeger](https://jaegertracing.io) implementation of the [OpenTracing](http://opentracing.io) API.
```
@jodh-intel one option would be to decouple the exporting from trace gathering. If we used open census we can export to any backend

https://opencensus.io/exporters/supported-exporters/go/

This will allow Kata to directly export traces to Cloud backends like Azure Monitor, AWS X-Ray (Tracing) and Stackdriver and Datadog.

This is a easy swap out.

Also open tracing has built in support for gRPC tracing

https://github.com/census-instrumentation/opencensus-go/blob/master/examples/grpc/helloworld_server/main.go#L62

jodh-intel replied:
It does sound like we have an embarrassment of riches when it comes to tracing options. 

I'll have to look at this next week. We already have otgrpc gRPC tracing support in the agent (today) but if you look at this PR, you'll see that I use that for "collated trace", but a custom gRPC tracer for "isolated tracing". So we could potentially switch the collated trace to use opencensus but that isn't "enough" for isolated tracing which traces the agent from startup to shutdown (not just grpc calls).

We may collectively decide that we only care about gRPC tracing. That certainly simplifies the problem significantly. However, I don't think it's ever going to be as useful as agent start-to-exit tracing as that will show what happened before the gRPC server came up for example.

jodh-intel commented:
For those keen to test this out, please note that until we've got agreement on how the VM is ended when tracing is enabled, you'll need to use the following runtime branch (which simply disables the VM killing call):

- https://github.com/jodh-intel/runtimes/tree/support-agent-tracing-only-disable-qmp-quit

mcastelino commented on scripts/vsock-to-udp-server.sh:
```diff
@@ -0,0 +1,49 @@
+export host_vsock_addr="x00x00xafx1ax00x00x02x00x00x00x00x00x00x00"
+
+# Create vsock server that redirects all traffic to UDP:6831 ("jaeger-agent")
+exec sudo socat -vx -u \
```
@jodh-intel why the `-vx` 

-v
Writes the transferred data not only to their target streams, but also to stderr. The output format is text with some conversions for readability, and prefixed with "> " or "< " indicating flow directions.
-x
Writes the transferred data not only to their target streams, but also to stderr. The output format is hexadecimal, prefixed with "> " or "< " indicating flow directions. Can be combined with -v .

jodh-intel replied:
Hi @mcastelino - I added so that you can see that Kata agent -> Jaeger agent comms are working.

mcastelino commented on agent.go:
```diff
@@ -24,12 +24,14 @@ import (
 	"time"
 
 	"github.com/gogo/protobuf/proto"
+	"github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc"
```
@jodh-intel otgrpc has `ocgrpc.DefaultServerViews` those capture all gRPC calls. If you use them you get a lot of the tracing for free at the gPRC level w/o instrumenting the actual callbacks.

mcastelino replied:
Sorry I meant ocgrpc has it. otgrpc does not seem to have an equivalent

mcastelino commented on tracing.go:
```diff
@@ -0,0 +1,105 @@
+
+// trace creates a new tracing span based on the specified contex, subsystem
+// and name.
+func trace(ctx context.Context, subsystem, name string) (opentracing.Span, context.Context) {
```
@jodh-intel this is a nice abstraction which almost completely decouples you from the actual underlying tracing implementation. However we return `opentracing.Span` here and then invoke the methods available in that object directly. e.g. finish()

We we provided methods or returned an interface here that we owned, then we could switch tracing frameworks w/o any code change anywhere else.

/cc @sboeuf

egernst replied:
This will be handled in a follow on PR, correct?  Is there an issue in place already which we can reference, @mcastelino  @sboeuf ?

jodh-intel replied:
Yep - the plan is I'll pass the baton to @mcastelino to switch the agent to opencensus once this lands.

mcastelino commented on agent.go:
```diff
@@ -786,21 +934,80 @@ func grpcTracer(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo
+// The reason this needs to return a function being that when this function is
+// called, the correct context to use for tracing is not available so the
+// returned function must query it when called later.
+func makeStreamInterceptor() grpc.StreamServerInterceptor {
```
@jodh-intel @sboeuf do we actually do streaming gRPC today.

jodh-intel replied:
I don't think so, but I added this to "cover all bases" ;)

mcastelino commented on agent.go:
```diff
@@ -524,6 +525,44 @@ func (s *sandbox) teardownSharedPidNs() error {
+	}
+
+	defer fieldLogger.Info("gRPC server stopped")
+
```
@sboeuf you can call finish of the global agent span here? The span does end a little bite early but we are done anyway at this point in time. So the global span will close before the last gRPC request closes, but we do not lose any real information.

amshinde commented on README.md:
```diff
@@ -15,3 +16,8 @@ The agent manages container processes inside the VM, on behalf of the
+## Developer mode
+
+Add `agent.devmode` to the guest kernel command line to enable
+[debug mode](#debug-mode) and allow the agent process to coredump (disabled by default).
```
Should this be developer mode?

jodh-intel commented:
Branch updated once again. Recent changes:

- Applied review feedback.
- Added `StartTracing()` and `StopTracing()` gRPC API's to support dynamic tracing.
- Updated the `TRACING.md` doc to explain dynamic vs. static tracing.

jodh-intel commented:
/cc @egernst, @mcastelino, @sboeuf.

jodh-intel commented:
This PR needs re-reviewing @sboeuf, @mcastelino, @egernst.

egernst commented:
travis has some compalints @jodh-intel:
```
/home/travis/gopath/src/github.com/kata-containers/agent/protocols/mockserver
protocols/mockserver/mockserver.go:54:38:warning: cannot use mock (variable of type *mockServer) as github.com/kata-containers/agent/protocols/grpc.AgentServiceServer value in argument to pb.RegisterAgentServiceServer: missing method StartTracing (staticcheck)
protocols/mockserver/mockserver.go:54:38:warning: unused struct field cannot use mock (variable of type *mockServer) as github.com/kata-containers/agent/protocols/grpc.AgentServiceServer value in argument to pb.RegisterAgentServiceServer: missing method StartTracing (structcheck)
protocols/mockserver/mockserver.go:54:38:warning: cannot use mock (variable of type *mockServer) as github.com/kata-containers/agent/protocols/grpc.AgentServiceServer value in argument to pb.RegisterAgentServiceServer: missing method StartTracing (maligned)
protocols/mockserver/mockserver.go:54:38:warning: unused variable or constant cannot use mock (variable of type *mockServer) as github.com/kata-containers/agent/protocols/grpc.AgentServiceServer value in argument to pb.RegisterAgentServiceServer: missing method StartTracing (varcheck)
protocols/mockserver/mockserver.go:54:38:warning: cannot use mock (variable of type *mockServer) as github.com/kata-containers/agent/protocols/grpc.AgentServiceServer value in argument to pb.RegisterAgentServiceServer: missing method StartTracing (unconvert)
protocols/mockserver/mockserver.go:54:38:warning: cannot use mock (variable of type *mockServer) as github.com/kata-containers/agent/protocols/grpc.AgentServiceServer value in argument to pb.RegisterAgentServiceServer: missing method StartTracing (unused)
```

jodh-intel commented:
I think it's mine. I'll fix whilst you guys continue to review.... ;)

jodh-intel commented:
@egernst - branch updated. Let's see what sort of mood Travis is in...

/retest

jodh-intel commented:
I should point out that the plan for the runtime is to call `StartTracing()` and `StopTracing()` when the runtime is configured for agent tracing. That will ensure the agent is *not* shut down when `DestroySandbox()` is called. We may add a special option in `configuration.toml` to allow the static agent tracing where `DestroySandbox()` would shut down the VM, but that will not be the default behaviour.

amshinde commented:
@jodh-intel CI is failing, failures are related to errors while deleting a container:
```
• Failure in Spec Teardown (AfterEach) [68.493 seconds]
state
/tmp/jenkins/workspace/kata-containers-agent-centos-7-4-PR/go/src/github.com/kata-containers/tests/functional/state_test.go:26
  container [AfterEach]
  /tmp/jenkins/workspace/kata-containers-agent-centos-7-4-PR/go/src/github.com/kata-containers/tests/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
    with workload [true], timeWait 5
    /tmp/jenkins/workspace/kata-containers-agent-centos-7-4-PR/go/src/github.com/kata-containers/tests/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:46

    Expected success, but got an error:
        <*errors.errorString | 0xc0001a8c00>: {
            s: "failed to delete container TBX1az8bZvgPWi5h79kI ",
        }
        failed to delete container TBX1az8bZvgPWi5h79kI 
```

jodh-intel commented:
@sboeuf - I think we just need you ack to allow us break out the champagne... :smile:

jodh-intel commented:
Gah - I'm starting to go off this browser with it's aggressive caching and page unloading - I'd missed @sboeuf's ack.

/me looks at test carnage...

jodh-intel commented:
/retest

jodh-intel commented:
16.04 metrics CI timed out @grahamwhaley:

```
Build timed out (after 10 minutes). Marking the build as aborted.
14:01:43 density/docker_memory_usage.sh: line 190: 29584 Terminated              docker rm -f ${containers[@]}
```

grahamwhaley commented:
@jodh-intel - that metrics CI timeout looks like docker/kata got stuck trying to delete the containers after the test had been run. It looks like @amshinde also saw a 'failed to delete container' on another CI before on this PR? That makes me wonder if we have some inherent rm/delete hangup hiding in here somewhere?
I'll re-trigger the CI (and I've tried to adjust the bounds, so the metrics CI has a chance of passing...)

jodh-intel commented:
Nooo - I've already re-triggered :)

Yes, there may be an issue but I was hoping to get a fresh set of CI logs as cannot recreate locally.

jodh-intel commented:
/retest

jodh-intel commented:
/retest

jodh-intel commented:
@nitkon - It looks like the ppc64 CI isn't actually running tests for PRs?:

```
Building remotely on Power8_slave01 (power8-ubuntu-1604 power8_node) in workspace /home/jenkins/workspace/kata-containers-agent-Power8-ubuntu-16-04-PR-initrd
[WS-CLEANUP] Deleting project workspace...
```

jodh-intel commented:
Branch updated.

@mcastelino - ftr, I removed the unused gRPC stream interceptor function.

sboeuf commented:
@mcastelino if you agree, let's merge this PR :)

mcastelino commented:
LGTM @sboeuf please merge this so that we can switch to a custom exporter quickly

sboeuf commented:
@jodh-intel that's it you're done, you can go to the beach for a year now :palm_tree: :sun_with_face:
